In [1]:
# Ejercicio 1: Pulsar Timing and Orbital Acceleration

# Info
h = 0.5  # time step (s)
t = [0.0, 0.5, 1.0, 1.5, 2.0]              # s
v = [200.0, 205.2, 210.8, 216.9, 223.5]    # km/s

# Task A: For Difference at t0
a_t0 = (v[1] - v[0]) / h

# Task B: Back Difference at t4
a_t4 = (v[4] - v[3]) / h

# Task C: Center Difference at t2
a_t2 = (v[3] - v[1]) / (2 * h)

print(f"Task A - a(t0 = {t[0]} s) = {a_t0:.2f} km/s^2")
print(f"Task B - a(t4 = {t[4]} s) = {a_t4:.2f} km/s^2")
print(f"Task C - a(t2 = {t[2]} s) = {a_t2:.2f} km/s^2")


Task A - a(t0 = 0.0 s) = 10.40 km/s^2
Task B - a(t4 = 2.0 s) = 13.20 km/s^2
Task C - a(t2 = 1.0 s) = 11.70 km/s^2


<div style="
background-color:#fff0f6;
border: 3px solid #ff9ac6;
padding:18px;
border-radius:20px;
box-shadow: 0 6px 15px rgba(255, 154, 198, 0.25);
font-family: Georgia;
">

<b>💖 Paso a paso</b><br><br>

1. Definí el paso h = 0.5 s y los datos.<br>
2. Apliqué diferencia hacia adelante en t0.<br>
3. Apliqué diferencia hacia atrás en t4.<br>
4. Apliqué diferencia centrada en t2.<br>
5. Mostré los resultados.

</div>


In [2]:
# Ejercicio 2: Stellar Structure and Density Gradients

# Info
r = [1.0, 1.1, 1.2, 1.3, 1.4]          # in 10^3 km
P = [10.00, 9.15, 8.20, 7.12, 5.90]    # arbitrary units
h = 0.1                                 # in 10^3 km

# Outermost point i = 4 -> use P4, P3, P2
P4 = P[4]
P3 = P[3]
P2 = P[2]

dP_dr_outer = (3*P4 - 4*P3 + P2) / (2*h)

print(f"dP/dr at r = {r[4]} (10^3 km) = {dP_dr_outer:.4f} (pressure units)/(10^3 km)")


dP/dr at r = 1.4 (10^3 km) = -12.9000 (pressure units)/(10^3 km)


<div style="border: 3px solid #ff7fb6; border-radius: 18px; padding: 16px 20px; background: #fff0f7;">

<h3 style="margin-top: 0; color: #c2185b;">🌸 Gradiente de Presión en el Borde</h3>

En el punto más externo \( r = 1.4 \) no podemos usar diferencia centrada porque no existe información fuera de la estrella.  
Por eso usamos la fórmula Backward de segundo orden:

$$
\frac{dP}{dr}\Big|_{i} \approx \frac{3P_i - 4P_{i-1} + P_{i-2}}{2h}
$$


Con:
- \( h = 0.1 \)
- \( P_4 = 5.90 \)
- \( P_3 = 7.12 \)
- \( P_2 = 8.20 \)

$$
\frac{dP}{dr}(1.4) = -12.90
$$

<hr style="border: none; border-top: 1px solid #ffb6d3;">

<b>🌷 Conclusión:</b>  


En la superficie de la estrella $r = 1.4$ obtuvimos

$$
\frac{dP}{dr} \approx -11.8
$$

Esto confirma que la presión disminuye hacia el exterior, tal como lo predice el modelo físico.

Usamos la fórmula backward de tres puntos porque estamos en el borde y no existe información más allá de la superficie.

Además, al ser un método de orden $O(h^2)$, es considerablemente más preciso:  
si reducimos $h$ por un factor de 10, el error se reduce aproximadamente por un factor de 100.

</div>



In [3]:
# Ejercicio 3: Galactic Dynamics and Redshift Gradients
# Calculamos dz/dD y d²z/dD² en el punto medio D = 14 Mpc

# Datos del ejercicio
D = [10, 12, 14, 16, 18]                     # Distancia (Mpc)
z = [0.0020, 0.0035, 0.0055, 0.0080, 0.0110] # Redshift
h = 2                                        # Paso en D (Mpc)

# En el punto medio (i=2, D=14) use:
# z1 = z_{i-1}, z2 = z_i, z3 = z_{i+1}
z1 = z[1]   # z(12)
z2 = z[2]   # z(14)
z3 = z[3]   # z(16)

# Fórmula centrada para primera derivada:
# dz/dD ≈ (z_{i+1} - z_{i-1}) / (2h)
dz_dD = (z3 - z1) / (2 * h)

# Fórmula centrada para segunda derivada:
# d²z/dD² ≈ (z_{i+1} - 2z_i + z_{i-1}) / h²
d2z_dD2 = (z3 - 2 * z2 + z1) / (h ** 2)

# Sustitución que pide el ejercicio
print("Sustitución para dz/dD:")
print(f"dz/dD ≈ (z3 - z1) / (2h)")
print(f"dz/dD ≈ ({z3:.4f} - {z1:.4f}) / (2*{h})")
print(f"dz/dD ≈ {dz_dD:.6f} 1/Mpc\n")

print("Sustitución para d²z/dD²:")
print(f"d²z/dD² ≈ (z3 - 2z2 + z1) / h²")
print(f"d²z/dD² ≈ ({z3:.4f} - 2*{z2:.4f} + {z1:.4f}) / ({h}²)")
print(f"d²z/dD² ≈ {d2z_dD2:.6f} 1/Mpc²")


Sustitución para dz/dD:
dz/dD ≈ (z3 - z1) / (2h)
dz/dD ≈ (0.0080 - 0.0035) / (2*2)
dz/dD ≈ 0.001125 1/Mpc

Sustitución para d²z/dD²:
d²z/dD² ≈ (z3 - 2z2 + z1) / h²
d²z/dD² ≈ (0.0080 - 2*0.0055 + 0.0035) / (2²)
d²z/dD² ≈ 0.000125 1/Mpc²


<div style="background-color:#ffe6f2; padding:20px; border-radius:15px; border:2px solid #ff66b2;">

## 🌸 Derivadas Centradas en el Punto Medio

En el punto medio $D = 14 \,\text{Mpc}$ usamos fórmulas centradas porque tenemos información a ambos lados del punto.

Datos:
- $z_1 = 0.0035$
- $z_2 = 0.0055$
- $z_3 = 0.0080$
- $h = 2 \,\text{Mpc}$

---

### 💖 Primera derivada

Fórmula centrada:

$$
\frac{dz}{dD} = \frac{z_{i+1} - z_{i-1}}{2h}
$$

Sustituyendo:

$$
\frac{dz}{dD} = \frac{0.0080 - 0.0035}{2(2)}
$$

$$
\frac{dz}{dD} = 0.001125 \,\text{Mpc}^{-1}
$$

---

### 💗 Segunda derivada

Fórmula centrada de segundo orden:

$$
\frac{d^2z}{dD^2} = \frac{z_{i+1} - 2z_i + z_{i-1}}{h^2}
$$

Sustituyendo:

$$
\frac{d^2z}{dD^2} = \frac{0.0080 - 2(0.0055) + 0.0035}{2^2}
$$

$$
\frac{d^2z}{dD^2} = 0.000125 \,\text{Mpc}^{-2}
$$

---

## 🌷 Conclusión

En $D=14\,\mathrm{Mpc}$ obtuvimos:

$$
\frac{dz}{dD}\approx 0.001125\ \mathrm{Mpc}^{-1},
\qquad
\frac{d^2z}{dD^2}\approx 0.000125\ \mathrm{Mpc}^{-2}.
$$

El hecho de que $\dfrac{dz}{dD}>0$ indica que el redshift $z$ aumenta con la distancia $D$, consistente con la expansión.

Usamos fórmulas **centradas** porque estamos en un punto interior (tenemos datos a ambos lados).  
Además, estas aproximaciones son de orden $O(h^2)$, por lo que el error de truncamiento decrece proporcionalmente a $h^2$, siendo más precisas que un esquema de primer orden $O(h)$.


</div>


In [8]:
# Ejercicio 5: Error Analysis and Consistency Discussion
# Pregunta 2: Convergence and Step Size (p = 2)
p = 2
h = 0.1
C = 1.0

E_h = C * h**p
E_half = C * (h/2)**p

factor = E_half / E_h

print(f"E(h)      = {E_h:.6f}")
print(f"E(h/2)    = {E_half:.6f}")
print(f"E(h/2)/E(h) = {factor:.6f}")
print(f"Reduction factor = {1/factor:.0f}  (i.e., error is 4x smaller)")






E(h)      = 0.010000
E(h/2)    = 0.002500
E(h/2)/E(h) = 0.250000
Reduction factor = 4  (i.e., error is 4x smaller)


<div style="background-color:#ffe6f0; padding:20px; border-radius:15px; border:2px solid #ffb6c1;">

<h1 style="color:#d63384; font-family:cursive;">💖 5. Error Analysis and Consistency Discussion</h1>

🩷 <b>① Theoretical Accuracy</b>  

El método centrado es más confiable que el forward porque su error de truncamiento es $O(h^2)$ y no $O(h)$.  
En el esquema centrado se cancelan términos de Taylor de primer orden, dejando como error dominante un término proporcional a $h^2$.

<hr style="border:1px solid #ffb6c1;">

🩷 <b>② Convergence and Step Size</b>  

Para un método de segundo orden ($p=2$), el error escala como:

$$
E(h) \sim h^2
$$

Si el paso se reduce a la mitad:

$$
E\left(\frac{h}{2}\right)
=
\left(\frac{1}{2}\right)^2 E(h)
=
\frac{1}{4}E(h)
$$

Por tanto, el error disminuye por un factor de 4.

<hr style="border:1px solid #ffb6c1;">

🩷 <b>③ High-Fidelity Simulations</b>  

En simulaciones astrofísicas de alta fidelidad, métodos $O(h)$ u $O(h^2)$ pueden ser insuficientes porque los errores por paso se acumulan en integraciones largas.  
Métodos de orden mayor (por ejemplo $O(h^4)$) reducen el error más rápido al refinar la malla, permitiendo mayor precisión y estabilidad.

</div>




